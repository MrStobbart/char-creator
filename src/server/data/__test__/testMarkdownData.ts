export const exampleRulesFile = `
This is some other Stuff

#### Soziale Talente

| *Talent* | *Anforderungen* | *Beschreibung* | *Wirkung* |
| ------------------------- | --------------------------- | ------------------------------------------------------------ | -------------------------- |
| *Demütigung* | Anfänger, Einschüchtern W8 | Solche mit einem herben Verstand können das Ego eines Gegners mit einer einzigen Anmerkung oder einer Geste zerstören. Der Charakter kann verspotten Proben wiederholen. | Provozieren wiederholbar |

#### Magische Talente

AH (Magier) impliziert auch Schamanen
`;

export const exampleTableExcerpt = `
#### Soziale Talente

| *Talent* | *Anforderungen* | *Beschreibung* | *Wirkung* |
| ------------------------- | --------------------------- | ------------------------------------------------------------ | -------------------------- |
| *Demütigung* | Anfänger, Einschüchtern W8 | Solche mit einem herben Verstand können das Ego eines Gegners mit einer einzigen Anmerkung oder einer Geste zerstören. Der Charakter kann verspotten Proben wiederholen. | Provozieren wiederholbar |

`;

export const exampleTableExcerpt2 = `
#### Fertigkeitsstufen

| Fertigkeitsstufe | Würfel | Bedeutung |
|--------------------|-------------|--------------|
| Nicht vorhanden | W4-1 | Ahnungslos |
| 1 | W4 | Anfänger |
| 2 | W6 | Ausgebildet |
| 3 | W8 | Kompetent |
| 4 | W10 | Herausragend |
| 5 | W12 | Profi |
| 6 | WildDie W8 | Legendär |
| 7 | WildDie W10 | |
| 8 | WildDie W12 | |`;

export const exampleTableWithLink = `
#### Magische Talente

AH (Magier) impliziert auch Schamanen

| *Talent* | *Anforderungen* | *Beschreibung* | *Wirkung* |
| ------------------------- | --------------------------- | ------------------------------------------------------------ | -------------------------- |
| *Initiation* | Fortgeschritten, AH, Zaubern W8, nur einmal pro Rang | Der Charakter erreicht die höheren Sphären der Magie seiner gewählten Profession. Er erhält jedes Mal wenn er dieses Talent wählt einen der folgenden Boni:<br />Ein kostenloses [Arkanes Talent der Initiation](#####Arkane-Talente-der-Initiation), jedes aber nur einmal<br />+1 auf Zaubern Proben (Magier)<br />Einen zusätzlichen Kraftpunkt (Adepten) | Verschiedenes |
| *Beschwörer* | Anfänger, AH(Magier), Beschwören W8 | Auf die Beschwörung von Geistern erhält der Charakter einen Bonus von +2 auf die Beschwören Probe. | +2 auf Bewschören |
| *Geisterpakt* | Veteran, Beschwörer, Beschwören W10 | Der Beschwörer hat einen Pakt mit der Geisterwelt abgeschlossen und kann nun mehrere Geister gleichzeitig kontrollieren. Sein neues Limit beträgt nun seine Willenskraft. | Doppel so viele Geister |
| *Ausdauernder Beschwörer* | Veteran, Beschwörer, Beschwören W8, Willenskraft W8 | Der Beschwörer erhält einen Bonus von +2 um dem Entzug von Beschwörungen zu Widerstehen. | +2 zum Entzug von Beschwören widerstehen |
| *Magische Ausdauer* | Fortgeschritten, Zaubern W8) | Der Charakter erhält einen Bonus von +1 zum Entzug widerstehen | +1 zum Entzug widerstehen |
| *Stärkere magische Ausdauer* | Veteran, Zaubern W10 | Der Charakter erhält einen Bonus von +2 zum Entzug widerstehen. | +2 zum Entzug widerstehen |
| *Magieradept* | Anfänger, AH(Adept) | Der Adept kann zusätzlich die Fertigkeit Zaubern erlernen, allerdings nicht über Fertigkeitspunkte, sondern er muss einen Ki-Punkt aufgeben um seine Stufe in Zaubern um einen Würfeltyp zu erhöhen (er muss diesen Wüfeltyp immer noch mit aufstiegen bezahlen. Der Adept erhält zusätzlich eine Macht die er nur über die Fertigkeit Zaubern wirken kann. Der Adept kann das Talent „Neue Mächte“ wählen. </br> Sollte dem Adepten Essenzverlust durch Bodytech oder einer anderen Quelle drohen, wird zuerst der Zaubern-Würfel davon beeinträchtigt. | Verschiedenes |
| *Neue Mächte* | Anfänger, Arkaner Hintergrund (Magier) | Ein arkaner Charaktertyp kann zwei neue Mächte erlernen, indem er dieses Talent auswählt. Er darf nur zwischen den Kräften wählen, die seinem arkanen Hintergrund normalerweise zur Verfügung stehen. Hier kann es je nach Setting Unterschiede geben. | 2 neue Mächte |
| *Reinigung* | Anfänger, AH(Magier) | Der Charakter kann Hintergrundstrahlung in einem Umkreis gleich seines Zauber-Würfels beseitigen. Der Vorgang ist ein Ritual das eine Stunde dauert. | Verschiedenes |
| *Verbesserte Kontrolle* | Fortgeschritten, Beschwörer | Der Charakter hat eine bessere Kontrolle über seine Geister. Entweder weil er sie gnadenlos dominiert oder weil er freundschaftlich mit ihnen umgeht. Daher wird der Modifikator für die Anzahl der Geister die er hält um 1 gesenkt. Das Talent kann mehrmals gewählt werden, aber nur einmal pro Rang. (Siehe Beschwören, Abschnitt *Kontrolle*) | Verschiedenes |
| Meisterhafter Magier | Anfänger, AH(Magier), Zaubern W8 | Dein Charakter kann den Entzug von einer Zauber um 1 erhöhen um die Ausprägung zu ändern. | Verschiedenes |

##### Arkane Talente der Initiation

| *Talent* | *Anforderungen* | *Beschreibung* | *Wirkung* |
| ------------------------- | -------------------`;
