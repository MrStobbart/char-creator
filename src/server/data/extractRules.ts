import { ruleLocations } from './ruleLocations';
import { config } from '../helpers/config';
import logger from '../helpers/winston';
import axios from 'axios';
import mdtable2json, { ParsedTable } from 'mdtable2json';

interface GitHubResponse {
  type: string;
  encoding: string;
  size: number;
  name: string;
  path: string;
  content: string;
  sha: string;
  url: string;
  git_url: string;
  html_url: string;
  download_url: string;
  _links: {
    git: string;
    self: string;
    html: string;
  };
}

interface GitHubErrorResponse {
  message: string;
}

const getRulesFromGithub = async (fileUrl: string) => {
  try {
    const response = await axios.get<GitHubResponse & GitHubErrorResponse>(fileUrl, {
      headers: {
        Authorization: `token ${config.githubToken}`,
      },
    });

    return Buffer.from(response.data.content, 'base64').toString();
  } catch (error) {
    logger.error(`Could not fetch and parse rule file from GitHub: ${error.message}`);
    throw error;
  }
};

export const getTableEndIndex = (ruleFile: string, startIndex: number): number => {
  const nextHashIndex = ruleFile.substring(startIndex).indexOf('\n#', 15);

  const fixedNextHashIndex = nextHashIndex === -1 ? ruleFile.length : nextHashIndex;

  if (ruleFile.charAt(fixedNextHashIndex - 1) === '(') {
    return getTableEndIndex(ruleFile, fixedNextHashIndex + 15);
  }

  return fixedNextHashIndex + startIndex;
};

export const parseTableForGivenHeader = (ruleFile: string, tableHeader: string): ParsedTable => {
  const tableHeaderIndex = ruleFile.indexOf(tableHeader);

  if (tableHeaderIndex === -1) throw new Error(`Table header '${tableHeader}' does not exist in rule file.`);

  const tableEndIndex = getTableEndIndex(ruleFile, tableHeaderIndex);

  const tableExcerpt = ruleFile.substring(tableHeaderIndex, tableEndIndex);
  const fixedTableExcerpt = tableExcerpt.replace(/[*]/g, '');

  return mdtable2json.getTables(fixedTableExcerpt)[0];
};

type TableData = { [subTitle: string]: ParsedTable };

const getRules = async (ruleSetName: string) => {
  const ruleLocation = ruleLocations[ruleSetName];
  if (!ruleLocation) throw new Error(`No rule set with the name ${ruleSetName} found`);

  const rules = await getRulesFromGithub(ruleLocation.fileUrl);

  const parsedData: { [title: string]: TableData } = {};
  ruleLocation.data.forEach(dataType => {
    const tableData: TableData = {};

    dataType.tableHeaders.forEach(tableHeader => {
      tableData[tableHeader] = parseTableForGivenHeader(rules, tableHeader);
    });

    parsedData[dataType.title] = tableData;
  });

  console.log(JSON.stringify(parsedData, null, 2));

  return parsedData;
};

getRules('savageRun');

/**
 * 1. Start string until next #
 * 2. Only if #does not have a ( in front of it
 */
