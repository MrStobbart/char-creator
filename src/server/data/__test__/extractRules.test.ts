import { exampleRulesFile, exampleTableExcerpt2, exampleTableWithLink } from './testMarkdownData';
import { getTableEndIndex, parseTableForGivenHeader } from '../extractRules';

describe('getTableEndIndex', () => {
  it('returns the correct endIndex of a table', () => {
    const lastTableIndex = getTableEndIndex(exampleRulesFile, 27);
    expect(lastTableIndex).toBe(505);
  });
  it.todo('can handle the special case with the (');
});

describe('parseTableForGivenHeader', () => {
  it('returns the object with the data from the table', () => {
    const parsedTable = parseTableForGivenHeader(exampleRulesFile, '#### Soziale Talente');
    expect(parsedTable).toMatchSnapshot();

    const parsedTable2 = parseTableForGivenHeader(exampleTableExcerpt2, '#### Fertigkeitsstufen');
    expect(parsedTable2).toMatchSnapshot();
  });

  it('parses the correct table when there is a link in the table', () => {
    const parsedTable = parseTableForGivenHeader(exampleTableWithLink, '#### Magische Talente');

    expect(parsedTable.headers.length).toBe(4);
    expect(parsedTable.json.length).toBe(11);

    expect(parsedTable).toMatchSnapshot();
  });
});
