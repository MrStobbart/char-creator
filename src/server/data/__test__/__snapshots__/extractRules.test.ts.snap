// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTableForGivenHeader parses the correct table when there is a link in the table 1`] = `
Object {
  "headers": Array [
    "Talent",
    "Anforderungen",
    "Beschreibung",
    "Wirkung",
  ],
  "json": Array [
    Object {
      "Anforderungen": "Fortgeschritten, AH, Zaubern W8, nur einmal pro Rang",
      "Beschreibung": " Der Charakter erreicht die höheren Sphären der Magie seiner gewählten Profession. Er erhält jedes Mal wenn er dieses Talent wählt einen der folgenden Boni:",
      "Talent": "Initiation",
      "Wirkung": "Verschiedenes",
    },
    Object {
      "Anforderungen": "Anfänger, AH(Magier), Beschwören W8",
      "Beschreibung": " Auf die Beschwörung von Geistern erhält der Charakter einen Bonus von +2 auf die Beschwören Probe.",
      "Talent": "Beschwörer",
      "Wirkung": "+2 auf Bewschören",
    },
    Object {
      "Anforderungen": "Veteran, Beschwörer, Beschwören W10",
      "Beschreibung": " Der Beschwörer hat einen Pakt mit der Geisterwelt abgeschlossen und kann nun mehrere Geister gleichzeitig kontrollieren. Sein neues Limit beträgt nun seine Willenskraft.",
      "Talent": "Geisterpakt",
      "Wirkung": "Doppel so viele Geister",
    },
    Object {
      "Anforderungen": "Veteran, Beschwörer, Beschwören W8, Willenskraft W8",
      "Beschreibung": " Der Beschwörer erhält einen Bonus von +2 um dem Entzug von Beschwörungen zu Widerstehen.",
      "Talent": "Ausdauernder Beschwörer",
      "Wirkung": "+2 zum Entzug von Beschwören widerstehen",
    },
    Object {
      "Anforderungen": "Fortgeschritten, Zaubern W8)",
      "Beschreibung": " Der Charakter erhält einen Bonus von +1 zum Entzug widerstehen",
      "Talent": "Magische Ausdauer",
      "Wirkung": "+1 zum Entzug widerstehen",
    },
    Object {
      "Anforderungen": "Veteran, Zaubern W10",
      "Beschreibung": " Der Charakter erhält einen Bonus von +2 zum Entzug widerstehen.",
      "Talent": "Stärkere magische Ausdauer",
      "Wirkung": "+2 zum Entzug widerstehen",
    },
    Object {
      "Anforderungen": "Anfänger, AH(Adept)",
      "Beschreibung": " Der Adept kann zusätzlich die Fertigkeit Zaubern erlernen, allerdings nicht über Fertigkeitspunkte, sondern er muss einen Ki-Punkt aufgeben um seine Stufe in Zaubern um einen Würfeltyp zu erhöhen (er muss diesen Wüfeltyp immer noch mit aufstiegen bezahlen. Der Adept erhält zusätzlich eine Macht die er nur über die Fertigkeit Zaubern wirken kann. Der Adept kann das Talent „Neue Mächte“ wählen. ",
      "Talent": "Magieradept",
      "Wirkung": "Verschiedenes",
    },
    Object {
      "Anforderungen": "Anfänger, Arkaner Hintergrund (Magier)",
      "Beschreibung": " Ein arkaner Charaktertyp kann zwei neue Mächte erlernen, indem er dieses Talent auswählt. Er darf nur zwischen den Kräften wählen, die seinem arkanen Hintergrund normalerweise zur Verfügung stehen. Hier kann es je nach Setting Unterschiede geben.",
      "Talent": "Neue Mächte",
      "Wirkung": "2 neue Mächte",
    },
    Object {
      "Anforderungen": "Anfänger, AH(Magier)",
      "Beschreibung": " Der Charakter kann Hintergrundstrahlung in einem Umkreis gleich seines Zauber-Würfels beseitigen. Der Vorgang ist ein Ritual das eine Stunde dauert.",
      "Talent": "Reinigung",
      "Wirkung": "Verschiedenes",
    },
    Object {
      "Anforderungen": "Fortgeschritten, Beschwörer",
      "Beschreibung": " Der Charakter hat eine bessere Kontrolle über seine Geister. Entweder weil er sie gnadenlos dominiert oder weil er freundschaftlich mit ihnen umgeht. Daher wird der Modifikator für die Anzahl der Geister die er hält um 1 gesenkt. Das Talent kann mehrmals gewählt werden, aber nur einmal pro Rang. (Siehe Beschwören, Abschnitt Kontrolle)",
      "Talent": "Verbesserte Kontrolle",
      "Wirkung": "Verschiedenes",
    },
    Object {
      "Anforderungen": "Anfänger, AH(Magier), Zaubern W8",
      "Beschreibung": " Dein Charakter kann den Entzug von einer Zauber um 1 erhöhen um die Ausprägung zu ändern.",
      "Talent": "Meisterhafter Magier",
      "Wirkung": "Verschiedenes",
    },
  ],
}
`;

exports[`parseTableForGivenHeader returns the object with the data from the table 1`] = `
Object {
  "headers": Array [
    "Talent",
    "Anforderungen",
    "Beschreibung",
    "Wirkung",
  ],
  "json": Array [
    Object {
      "Anforderungen": "Anfänger, Einschüchtern W8",
      "Beschreibung": "Solche mit einem herben Verstand können das Ego eines Gegners mit einer einzigen Anmerkung oder einer Geste zerstören. Der Charakter kann verspotten Proben wiederholen.",
      "Talent": "Demütigung",
      "Wirkung": "Provozieren wiederholbar",
    },
  ],
}
`;

exports[`parseTableForGivenHeader returns the object with the data from the table 2`] = `
Object {
  "headers": Array [
    "Fertigkeitsstufe",
    "Würfel",
    "Bedeutung",
  ],
  "json": Array [
    Object {
      "Bedeutung": "Ahnungslos",
      "Fertigkeitsstufe": "Nicht vorhanden",
      "Würfel": "W4-1",
    },
    Object {
      "Bedeutung": "Anfänger",
      "Fertigkeitsstufe": "1",
      "Würfel": "W4",
    },
    Object {
      "Bedeutung": "Ausgebildet",
      "Fertigkeitsstufe": "2",
      "Würfel": "W6",
    },
    Object {
      "Bedeutung": "Kompetent",
      "Fertigkeitsstufe": "3",
      "Würfel": "W8",
    },
    Object {
      "Bedeutung": "Herausragend",
      "Fertigkeitsstufe": "4",
      "Würfel": "W10",
    },
    Object {
      "Bedeutung": "Profi",
      "Fertigkeitsstufe": "5",
      "Würfel": "W12",
    },
    Object {
      "Bedeutung": "Legendär",
      "Fertigkeitsstufe": "6",
      "Würfel": "WildDie W8",
    },
    Object {
      "Bedeutung": "",
      "Fertigkeitsstufe": "7",
      "Würfel": "WildDie W10",
    },
    Object {
      "Bedeutung": "",
      "Fertigkeitsstufe": "8",
      "Würfel": "WildDie W12",
    },
  ],
}
`;
